types/wlr_layer_shell.c

```diff
@@ -435,3 +448,74 @@ void wlr_layer_shell_destroy(struct wlr_layer_shell *layer_shell) {
 	wl_global_destroy(layer_shell->wl_global);
 	free(layer_shell);
 }
+
+struct layer_surface_iterator_data {
+	wlr_surface_iterator_func_t user_iterator;
+	void *user_data;
+	int x, y;
+};
+
+static void layer_surface_iterator(struct wlr_surface *surface,
+		int sx, int sy, void *data) {
+	struct layer_surface_iterator_data *iter_data = data;
+	iter_data->user_iterator(surface, iter_data->x + sx, iter_data->y + sy,
+		iter_data->user_data);
+}
+
+static void xdg_surface_for_each_surface(struct wlr_xdg_surface *surface,
+		int x, int y, wlr_surface_iterator_func_t iterator, void *user_data) {
```

Comment:
You can use `wlr_xdg_surface_for_each_surface` instead of copy-pasting it here.

Reply:
I had issues with that

Reply:
It shouldn't be hard, you can use the `data` struct you defined earlier

Reply:
That wasn't the issue. It was related to double-rendering popups